chain/evm/compass.go

```diff
@@ -617,3 +840,111 @@ func (c compass) callCompass(
 	}
 	return c.evm.ExecuteSmartContract(ctx, c.chainID, *c.compassAbi, c.smartContractAddr, useMevRelay, method, arguments)
 }
+
+func (t compass) gravityRelayBatches(ctx context.Context, batches []chain.GravityBatchWithSignatures) error {
+	var gErr whoops.Group
+	logger := log.WithField("chainReferenceID", t.ChainReferenceID)
+	for _, batch := range batches {
+		logger = logger.WithField("batch-nonce", batch.BatchNonce)
+
+		if ctx.Err() != nil {
+			logger.Debug("exiting processing batch context")
+			break
+		}
+
+		var processingErr error
+		var tx *ethtypes.Transaction
+		logger := logger.WithFields(log.Fields{
+			"batch-nonce": batch.BatchNonce,
+		})
+		logger.Debug("relaying")
+
+		tx, processingErr = t.gravityRelayBatch(ctx, batch)
+
+		processingErr = whoops.Enrich(
+			processingErr,
+			FieldMessageID.Val(batch.BatchNonce),
+		)
+
+		switch {
+		case processingErr == nil:
+			if tx != nil {
+				logger.Debug("sending claim")
+				// TODO : Claim
+				//if err := t.paloma.SetPublicAccessData(ctx, t.ChainReferenceID, batch.BatchNonce, tx.Hash().Bytes()); err != nil {
+				//	gErr.Add(err)
+				//	return gErr
+				//}
+			}
+		case goerrors.Is(processingErr, ErrNoConsensus):
+			// does nothing
+		default:
+			logger.WithError(processingErr).Error("relay error")
+			gErr.Add(processingErr)
+		}
+	}
+
+	return gErr.Return()
+}
+
+func (t compass) gravityRelayBatch(
+	ctx context.Context,
+	batch chain.GravityBatchWithSignatures,
+) (*ethtypes.Transaction, error) {
+	return whoops.TryVal(func() *ethtypes.Transaction {
+		executed, err := t.gravityIsBatchAlreadyRelayed(ctx, batch.BatchNonce)
+		whoops.Assert(err)
+		if executed {
+			return nil
+		}
+
+		valsetID, err := t.findLastValsetMessageID(ctx)
+		whoops.Assert(err)
+
+		valset, err := t.paloma.QueryGetEVMValsetByID(ctx, valsetID, t.ChainReferenceID)
+		whoops.Assert(err)
+
+		consensusReached := isConsensusReached(ctx, valset, batch)
+		if !consensusReached {
+			whoops.Assert(ErrNoConsensus)
+		}
+
+		con := BuildCompassConsensus(ctx, valset, batch.Signatures)
+
+		receivers := make([]common.Address, len(batch.Transactions))
+		amounts := make([]*big.Int, len(batch.Transactions))
+
+		for i, transaction := range batch.Transactions {
+			receivers[i] = common.HexToAddress(transaction.DestAddress)
+			amounts[i] = transaction.Erc20Token.Amount.BigInt()
+		}
+
+		compassArgs := CompassTokenSendArgs{
+			Receiver: receivers,
+			Amount:   amounts,
+		}
+
+		tx, err := t.callCompass(
+			ctx,
+			false,
+			"submit_batch",
+			[]any{
+				con,
+				common.HexToAddress(batch.TokenContract),
+				compassArgs,
+				new(big.Int).SetInt64(int64(batch.BatchNonce)),
+				new(big.Int).SetInt64(int64(batch.GetBatchTimeout())), // TODO : Deadline
+			},
+		)
+		if err != nil {
+			// TODO : Where to store data on error?
+			//isSmartContractError := whoops.Must(t.SetErrorData(ctx, batch.BatchNonce, err))
+			//if isSmartContractError {
+			//	return nil
+			//}
```

Comment:
Is it not possible to use the same logic used or `submit_logic_call`?

Reply:
We can, it's just not implemented.  Gravity has its own stores.  It's not going through the evm queues.  I would ask that we do this afterward and not block this PR if that's OK.  I'll update this code to output a log message, so at least the error is somewhere.

Reply:
Yeah that's absolutely fine. As long as there's a ticket or something that reminds us this is outstanding. I've come across TODOs from over a year ago in the code, I don't trust relying on them alone ;)

Reply:
Agreed.  Code is a terrible ticket tracker.  I've added it here: https://www.notion.so/volumefi/Pigeon-Paloma-tech-debt-tally-bf0ee61cb76241719a39761c9577093e